core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,226 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.executor
+
+import java.io._
+import java.nio.charset.Charset
+import java.nio.file.{Files, Paths}
+import java.util.Locale
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.{SparkEnv, SparkException}
+import org.apache.spark.internal.{config, Logging}
+import org.apache.spark.util.Utils
+
+private[spark] case class ProcfsBasedSystemsMetrics(
+    jvmVmemTotal: Long,
+    jvmRSSTotal: Long,
+    pythonVmemTotal: Long,
+    pythonRSSTotal: Long,
+    otherVmemTotal: Long,
+    otherRSSTotal: Long)
+
+// Some of the ideas here are taken from the ProcfsBasedProcessTree class in hadoop
+// project.
+private[spark] class ProcfsBasedSystems(val procfsDir: String = "/proc/") extends Logging {
+  val procfsStatFile = "stat"
+  val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+  var pageSize = computePageSize()
+  var isAvailable: Boolean = isProcfsAvailable
+  private val pid = computePid()
+  private val ptree = mutable.Map[ Int, Set[Int]]()
+
+  var allMetrics: ProcfsBasedSystemsMetrics = ProcfsBasedSystemsMetrics(0, 0, 0, 0, 0, 0)
+
+  computeProcessTree()
+
+  private def isProcfsAvailable: Boolean = {
+    if (testing) {
+      return true
+    }
+    try {
+      if (!Files.exists(Paths.get(procfsDir))) {
+        return false
+      }
+    }
+    catch {
+      case f: FileNotFoundException => return false
+    }
+    val shouldLogStageExecutorMetrics =
+      SparkEnv.get.conf.get(config.EVENT_LOG_STAGE_EXECUTOR_METRICS)
+    val shouldLogStageExecutorProcessTreeMetrics =
+      SparkEnv.get.conf.get(config.EVENT_LOG_PROCESS_TREE_METRICS)
+    shouldLogStageExecutorProcessTreeMetrics && shouldLogStageExecutorMetrics
+  }
+
+  private def computePid(): Int = {
+    if (!isAvailable || testing) {
+      return -1;
+    }
+    try {
+      // This can be simplified in java9:
+      // https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.html
+      val cmd = Array("bash", "-c", "echo $PPID")
+      val length = 10
+      val out2 = Utils.executeAndGetOutput(cmd)
+      val pid = Integer.parseInt(out2.split("\n")(0))
+      return pid;
+    }
+    catch {
+      case e: SparkException => logDebug("IO Exception when trying to compute process tree." +
```

Comment:
why only SparkException, not any Exception?  also the msg shouldn't say "IO Exception".

should probably be logWarn

Reply:
Let me double check I thought there was a comment before that I should just get SparkException, but you are right. it doesn't make sense. Probably a mistake on my side. I was just caring about IOException here.

Reply:
well, executeAndGetOutput might throw a SparkException ... but are you sure nothing else will get thrown?  Eg. what if you get some weird output and then the `Integer.parseInt` failse?  Is there some reason you wouldn't want the same error handling for any exception here?

Reply:
At first I was getting all throwables. Then I thought it can be dangerous. There was also a review comment about that. So Not sure what is the correct way of handling this. Is it better to just take care of exceptions that we know can be thrown or catch all throwables?

Reply:
there's a distinction between `Throwable` and `Exception` -- `Throwable` includes Errors which are fatal to the JVM, you probably can't do anything.

In general its a good question whether you should catch specific exceptions or everything.  Here, you're calling an external program, and I don't feel super confident that we know how it always behaves, so I think we should be a little extra cautious.  An unhandled exception here would lead to not sending any heartbeats, which would be really bad.  Except for JVM errors, I think we just want to turn off this particular metric and keep going.